rust/wal3/src/lib.rs

```diff
@@ -0,0 +1,667 @@
+        ExponentialBackoff::new(options.throughput as f64, options.headroom as f64)
+    }
+}
+
+///////////////////////////////////////// LogWriterOptions /////////////////////////////////////////
+
+#[derive(Clone, Eq, PartialEq, arrrg_derive::CommandLine)]
+pub struct LogWriterOptions {
+    #[arrrg(nested)]
+    pub format: LogOptions,
+    #[arrrg(optional, "Number of shards in the log.")]
+    pub shards: usize,
+    #[arrrg(nested)]
+    pub throttle_shard: ThrottleOptions,
+    #[arrrg(nested)]
+    pub throttle_manifest: ThrottleOptions,
+    #[arrrg(optional, "The alpha for manifest load.")]
+    pub load_alpha: usize,
+    #[arrrg(optional, "The alpha for manifest store.")]
+    pub store_alpha: usize,
+}
+
+impl Default for LogWriterOptions {
+    fn default() -> Self {
+        LogWriterOptions {
+            // Default log options.
+            format: LogOptions::default(),
+            // Start with a log of one shard.
+            shards: 1,
+            // Default throttling options for shards.
+            throttle_shard: ThrottleOptions::default(),
+            throttle_manifest: ThrottleOptions::default(),
+            // Default alpha for manifest load.
+            load_alpha: 1,
+            // Default alpha for manifest store.
+            store_alpha: 1,
+        }
+    }
+}
+
+///////////////////////////////////////////// LogWriter ////////////////////////////////////////////
+
+pub struct LogWriter<O: ObjectStore> {
+    /// LogWriterOptions are fixed at log creation time.
+    /// LogWriter is intentionally cheap to construct and destroy.
+    /// Reopen the log to change the options.
+    options: Arc<LogWriterOptions>,
+    /// A generic object store.
+    object_store: Arc<O>,
+    /// True iff the log is done.
+    done: AtomicBool,
+    /// StreamManager manages the open streams on this log.
+    stream_manager: StreamManager,
+    /// ManifestManager coordinates updates to the manifest.
+    manifest_manager: ManifestManager,
+    /// ShardManager coordinates writes to the non-manifest shards.
+    shard_manager: ShardManager,
+    /// A channel to keep work alive on.
+    #[allow(dead_code)]
+    reap: tokio::sync::mpsc::Sender<tokio::task::JoinHandle<()>>,
+    /// A background future that collects dead tasks
+    reaper: Mutex<Option<tokio::task::JoinHandle<()>>>,
+    /// A background future that flushes the log.
+    flusher: Mutex<Option<tokio::task::JoinHandle<()>>>,
+}
+
+impl<O: ObjectStore> LogWriter<O> {
+    pub async fn initialize(options: &LogWriterOptions, object_store: &O) -> Result<(), Error> {
+        Manifest::initialize(options, object_store).await
+    }
+
+    pub async fn open(mut options: LogWriterOptions, object_store: O) -> Result<Arc<Self>, Error> {
+        options.throttle_manifest.outstanding =
+            std::cmp::min(options.throttle_manifest.outstanding, options.store_alpha);
+        options.store_alpha = options.throttle_manifest.outstanding;
```

Comment:
nit: please leave a comment on why `outstanding` and `store_alpha` can be set to the min of the two